Guard against unset resolved_archive_file

dmlap opened the pull request:
# What does this PR do?

`resolved_archive_file` in `_load_pretrained_model()` appears to be optional. In my case, I was loading a model from a GGUF file and it was `None`:

```python
model = AutoModelForCausalLM.from_pretrained(train_config.model_name,
                                             device_map='auto',
                                             gguf_file='llama3.2.gguf',
                                             offload_folder='offload')
```

In that case, `archive_file` ends up being `None` and the check for safe tensors raises an error. The change guards against that case and allows loading to continue.

I thought this change was minor enough that new tests were not warranted. If you feel otherwise, happy to add one if you can point me at the right place to do it.

dmlap committed:
archive_file may not be specified
When loading a pre-trained model from a gguf file, resolved_archive_file may not be set. Guard against that case in the safetensors availability check.

SunMarc reviewed (approved):
LGTM ! This is quite an edge case as we have `resolved_archive_file = None` when loading with gguf + we have `disk` in `device_map`. If you have time, please add a test in the tests/quantization/ggml/test_ggml.py file.  cc @Isotr0py for visibility

Isotr0py reviewed (approved):
LGTM too! Just need a test case to cover this edge case in `test_ggml.py`.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35628). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

dmlap commented:
Great! I should be able to add one sometime this week

SunMarc commented:
After reflection, I think that we shouldn't allow offload with GGUF. This is because with gguf state_dict, we still have to modify the state dict to be compatible with transformers. So we can't really offload to disk.

dmlap commented:
That makes sense. Do you think that should happen in transformers or accelerate? If it’s in the model loading here, I don’t mind taking a crack at it.

SunMarc commented:
The check should be in transformers !

dmlap committed:
Remap partial disk offload to cpu for GGUF files
GGUF files don't support disk offload so attempt to remap them to the CPU when device_map is auto. If device_map is anything else but None, raise a NotImplementedError.

dmlap commented:
@SunMarc I modified the handling of `device_map` so when `”auto”` is specified for a GGUF file, it will attempt to remap disk offload back to the CPU. If disk is explicitly configured, a `NotImplementedError` will be raised.

There’s a test case for the explicit disk mapping but there didn’t seem to be a non-invasive way of testing the `auto` remapping, and it didn’t seem worth it to me. Let me know if you disagree or would like any further modifications.

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -4218,12 +4218,31 @@ def from_pretrained(
             if hf_quantizer is not None:
                 hf_quantizer.validate_environment(device_map=device_map)
 
+            if gguf_path:
+                remapped_devices = set()
+                for name, device in device_map.items():
+                    if device == "disk":
+                        device_map[name] = "cpu"
+                        remapped_devices.add(name)
+                if len(remapped_devices) > 0:
+                    logger.warning(
+                        "Accelerate has auto-mapped modules to disk but disk offload is not supported for "
+                        "models loaded from GGUF files. Remapping modules to the cpu: "
+                        ", ".join(remapped_devices)
+                    )
+
```
I prefer not remap the device_map but raise an error afterwards.

dmlap replied:
I did this remapping because I originally hit this issue trying to run a GGUF on my MacBook with `device_map=“auto”` and got that confusing error with the unspecified archive file. If the disk modules had been mapped to cpu instead, everything should have worked.

Now that I’ve dug in quite a bit more, an explicit device mapping (like in the test case) would get past my issue but it does raise the sophistication requirements for developer-users quite a bit. That is, as a user I would have preferred if transformers would have remapped my model or at least warned me.

All that said, let me know if you still think the remapping should be removed and I’ll pull it.

SunMarc replied:
>  If the disk modules had been mapped to cpu instead, everything should have worked.
With the infer_auto_device_map, we map to disk if we don't have any space left on the other devices. So mapping disk to cpu could cause an issue. I prefer just raising a warning saying that disk offload with gguf is not supported.

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -4218,12 +4218,31 @@ def from_pretrained(
             check_tied_parameters_on_same_device(tied_params, device_map)
 
+        if gguf_path and device_map is not None and "disk" in device_map.values():
+            raise NotImplementedError(
+                "One or more modules is configured to be mapped to disk. Disk offload is not supported for models "
+                "loaded from GGUF files."
+            )
```
Let's just do `RuntimeError` instead of `NotImplementedError`

SunMarc reviewed (commented):
Thanks for iterating, just a few nits

dmlap committed:
Don't remap auto device_map and raise RuntimeError
If device_map=auto and modules are selected for disk offload, don't attempt to map them to any other device. Raise a runtime error when a GGUF model is configured to map any modules to disk.

dmlap commented:
@SunMarc ok, I think all the comments are addressed. Let me know if any additional changes are needed

SunMarc committed:
Merge branch 'main' into no-archive-no-safetensors

SunMarc reviewed (approved):
Thanks for the discussion and iterating !

dmlap commented:
No problem! Thanks for the feedback
